// @haelp/teto/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get CH () {
        return _channel.CH;
    },
    get ChannelAPI () {
        return _channel.ChannelAPI;
    },
    get Classes () {
        return _classes;
    },
    get Client () {
        return _classes.Client;
    },
    get Engine () {
        return _engine;
    },
    get Types () {
        return _types;
    },
    get Utils () {
        return _utils;
    },
    get ch () {
        return _channel.ch;
    },
    get version () {
        return _utils.version;
    }
});
const _utils = /*#__PURE__*/ _interop_require_wildcard(require("./utils"));
const _chalk = /*#__PURE__*/ _interop_require_default(require("chalk"));
const _channel = require("./channel");
const _types = /*#__PURE__*/ _interop_require_wildcard(require("./types"));
const _classes = /*#__PURE__*/ _interop_require_wildcard(require("./classes"));
const _engine = /*#__PURE__*/ _interop_require_wildcard(require("./engine"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const suppressKey = "TRIANGLE_VERSION_SUPPRESS";
if (typeof process !== "undefined" && !(suppressKey in process.env)) {
    fetch("https://registry.npmjs.org/@haelp/teto").then((r)=>r.json()).then((d)=>{
        if (_utils.version < d["dist-tags"].latest) console.log(`${_chalk.default.redBright("[Triangle.js]")} Your triangle.js is out of date (v${_utils.version} vs v${d["dist-tags"].latest}). We recommend updating with 'npm install @haelp/teto@latest'.\nTo suppress this warning, set the ${suppressKey} environment variable.`);
    });
}
